hw/top_earlgrey/data/top_earlgrey.hjson

```diff
@@ -8,11 +8,51 @@
 
   datawidth: "32",  # 32-bit datawidth
 
-  clocks: [
-    { name: "main", freq: "100000000" }
-    { name: "fixed", freq: "100000000" }
-    { name: "usb", freq: "48000000" }
-  ]
+  // This is the clock data strcture of the design.
+  // The src key indicates the raw clock sources in the design
+  // The groups key indicates the various clock groupings in the design
+  clocks: {
+
+    srcs: [
+      { name: "main",  freq: "100000000" }
+      { name: "fixed", freq: "100000000" }
+      { name: "usb",   freq: "48000000" }
+    ],
+
+    // Clock Group attributes
+    // name: name of group.
+    //
+    // sw_cfg: whether software is allowed to gate the clock
+    // "no"   - software is not allowed to gate clocks
+    // "yes"  - software is allowed to gate clocks
+    // "hint" - software can provide a hint, and hw controls the rest
+    //
+    // unique: whether each module in the group can be separately gated
+    //         if sw_cfg is "no", this field has no meaning
+    // "yes"  - each clock is individually controlled
+    // "no"   - the group is controlled as one single unit
+    //
+    // The powerup and proc groups are unique.
+    // The powerup group of clocks do not need through the clock
+    // controller as they manage clock controller behavior
+    // The proc group is not peripheral, and direclty hardwired
+
+    groups: [
+      { name: "powerup", sw_cg: "no"                   }
+      { name: "trans",   sw_cg: "hint", unique: "yes", }
+      { name: "infra",   sw_cg: "no",                  }
+      { name: "secure",  sw_cg: "no"                   }
+      { name: "peri",    sw_cg: "yes",  unique: "no"   }
+      { name: "timers",  sw_cg: "no"                   }
+      { name: "proc",
```

Comment:
@eunchan 
notice how the process group is directly hardcoded. I couldn't think of another good way to make use of the same infrastructure without doing this.

Reply:
i probably should add a helper function that can be used to pull the clock from the template file.

Reply:
I agree. Unless the tool parses which module uses which clock and where the module sits in the clock domain, it will be hard to connect the clock at top to the module.